packages/workspace-resolver/src/resolve-workspace.ts

```diff
@@ -0,0 +1,187 @@
+import path, { join } from 'path';
+import { readJson } from 'fs-extra';
+import globby from 'globby';
+import semver from 'semver';
+
+import type {
+  ModularWorkspacePackage,
+  ModularType,
+  WorkspaceMap,
+  WorkspaceObj,
+} from 'modular-types';
+
+function packageJsonPath(dir: string) {
+  return dir.endsWith('package.json') ? dir : join(dir, 'package.json');
+}
+
+function resolveWorkspacesDefinition(
+  cwd: string,
+  def: PackageJson['workspaces'],
+): string[] {
+  if (!def) {
+    return [];
+  }
+
+  if (Array.isArray(def)) {
+    return def.flatMap((path: string) => {
+      return globby.sync(
+        [`${path}/package.json`, '!**/node_modules/**/*', '!**/__tests__/**/*'],
+        {
+          absolute: false,
+          cwd,
+        },
+      );
+    });
+  }
+
+  return resolveWorkspacesDefinition(cwd, def.packages);
+}
+
+type PackageJson = {
+  name: string;
+  version: string;
+  workspaces?: string[] | { noHost: boolean; packages: string[] };
+  modular?: { type: ModularType };
+  optionalDependencies: Record<string, string> | undefined;
+  devDependencies: Record<string, string> | undefined;
+  dependencies: Record<string, string> | undefined;
+};
+
+function readPackageJson(
+  isRoot: boolean,
+  workingDir: string,
+  relativePath: string,
+): Promise<PackageJson> {
+  if (isRoot) {
+    return readJson(relativePath) as Promise<PackageJson>;
```

Comment:
nit: instead of rewriting the whole `readJson` signature twice, can we just conditionally create the filename that we're reading?

Reply:
👍 shame `readJson` doesn't allow for generics, or supplying a schema to validate the JSON while it's being read...